compiler/pkg/openapi_generator/check-backward-compatibility.go

```diff
@@ -0,0 +1,142 @@
+		if err != nil {
+			return nil, fmt.Errorf("error unmarshaling new CRD: %v", err)
+		}
+
+		if newCRD.Name != existingCRDName {
+			continue
+		}
+
+		// When there is a backward incompatibility, we fail the build if we don't force an upgrade.
+		isInCompatible, message, err := nexus_compare.CompareFiles([]byte(existingCRDContent), []byte(newCRDPart))
+		if err != nil {
+			panic(fmt.Sprintf("Error occurred while checking CRD's %q backward compatibility: %v", existingCRDName, err))
+		}
+		if isInCompatible {
+			log.Warnf("CRD %q is incompatible with the previous version", existingCRDName)
+			inCompatibleCRDs = append(inCompatibleCRDs, message)
+		}
+	}
+	return inCompatibleCRDs, nil
+}
+
+func CheckBackwardCompatibility(existingCRDsPath, yamlsPath string, force bool) error {
+	var (
+		removedCRDs      []string
+		inCompatibleCRDs []*bytes.Buffer
+	)
+
+	if err := filepath.Walk(existingCRDsPath, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			if errors.Is(err, os.ErrNotExist) {
+				fmt.Printf("No files exists on the path %q: %v", path, err)
+				return nil
+			}
+			return fmt.Errorf("walking existing CRD files: %v", err)
+		}
+
+		if info.IsDir() {
+			fmt.Printf("Skipping dir %q\n", path)
+			return nil
+		}
+
+		if !strings.HasSuffix(info.Name(), ".yaml") {
+			return nil
+		}
+
+		existingCRDContent, err := os.ReadFile(path)
+		if err != nil {
+			return fmt.Errorf("error reading the existing CRD file on path %q: %v", path, err)
+		}
+
+		/*
+			Find the new CRD for the file with the same name as the existing CRD file and compare it to the new CRD spec provided.
+				existingCRDsPath - indicates the directory path of the existing crd yamls
+					Ex: the path will be `example/output/generated/crds`
+
+				path - indicates the file path of node
+					Ex: For the node `Config` the path will be `example/output/generated/crds/config_config.yaml`
+
+				yamlsPath - indicates the directory path of new crd yamls
+					Ex: the path will be `_generated/crds`
+		*/
+
+		newFilePath := yamlsPath + strings.TrimPrefix(path, existingCRDsPath)
+		newCRDContent, err := os.ReadFile(newFilePath)
+		if err != nil && !errors.Is(err, os.ErrNotExist) {
+			return fmt.Errorf("error reading the crd file on the path %q: %v", newFilePath, err)
+		}
+
+		existingCRDParts := splitCRDs(existingCRDContent)
+		for _, existingCRDPart := range existingCRDParts {
+			if existingCRDPart == "" {
+				continue
+			}
+
+			existingCRD := &extensionsv1.CustomResourceDefinition{}
+			err := yaml.Unmarshal([]byte(existingCRDPart), existingCRD)
+			if err != nil {
+				return fmt.Errorf("error unmarshaling existing CRD: %v", err)
+			}
+
+			// Appears node is removed in the latest version
+			if len(newCRDContent) == 0 {
+				removedCRDs = append(removedCRDs, existingCRD.Name)
+				continue
+			}
+
+			if inCompatibleCRDs, err = compareCRDs(inCompatibleCRDs, existingCRD.Name, existingCRDPart, newCRDContent); err != nil {
```

Comment:
Looks like we are anticipating that existingCRD file may have multiple CRD's.
If yes, then new files can also have multiple CRD's correct ?